src/cmap/wire_protocol/on_demand/document.ts

```diff
@@ -0,0 +1,320 @@
+import {
+  Binary,
+  BSON,
+  type BSONElement,
+  BSONError,
+  type BSONSerializeOptions,
+  BSONType,
+  getBigInt64LE,
+  getFloat64LE,
+  getInt32LE,
+  ObjectId,
+  parseToElementsToArray,
+  Timestamp,
+  toUTF8
+} from '../../../bson';
+
+// eslint-disable-next-line no-restricted-syntax
+const enum BSONElementOffset {
+  type = 0,
+  nameOffset = 1,
+  nameLength = 2,
+  offset = 3,
+  length = 4
+}
+
+export type JSTypeOf = {
+  [BSONType.null]: null;
+  [BSONType.undefined]: null;
+  [BSONType.double]: number;
```

Comment:
I'd expect this to be `Double`, and then `getNumber` / `getFlexibleNumber` would coerce the value to a `number`.  But right now, there's no way to decode a BSON double into a boxed type

Reply:
I would expect the opposite since the boxed double type isn't useful to the driver. and we _only_ need double for one case which is "ok" so sending it right to a JS number skips a step we will always perform anyway.

Reply:
This is the only type in this list that has an implicit, lossy conversion.  It feels wrong to have one special case here.  Isn't the purpose of getFlexibleNumber to provide a wrapper that the driver can use that doesn't care about numeric box types?

Reply:
from slack: this approach still maintains precision